src/Streaming/Internal.hs

```diff
@@ -136,46 +138,47 @@ deriving instance (Eq r, Eq (m (Stream f m r))
 deriving instance (Typeable f, Typeable m, Data r, Data (m (Stream f m r))
                   , Data (f (Stream f m r))) => Data (Stream f m r)
 #endif
-instance (Functor f, Monad m) => Functor (Stream f m) where
+-}
+instance Monad m => Functor (Stream f m) where
   fmap f = loop where
     loop stream = case stream of
       Return r -> Return (f r)
       Effect m  -> Effect (do {stream' <- m; return (loop stream')})
-      Step f   -> Step (fmap loop f)
+      Step f k  -> Step f (loop . k)
   {-# INLINABLE fmap #-}
   a <$ stream0 = loop stream0 where
     loop stream = case stream of
       Return r -> Return a
       Effect m  -> Effect (do {stream' <- m; return (loop stream')})
-      Step f    -> Step (fmap loop f)
+      Step f k  -> Step f (loop . k)
   {-# INLINABLE (<$) #-}  
 
-instance (Functor f, Monad m) => Monad (Stream f m) where
+instance Monad m => Monad (Stream f m) where
   return = Return
   {-# INLINE return #-}
   stream1 >> stream2 = loop stream1 where
     loop stream = case stream of
       Return _ -> stream2
       Effect m  -> Effect (liftM loop m)
-      Step f   -> Step (fmap loop f)  
+      Step f k  -> Step f (loop . k)
   {-# INLINABLE (>>) #-}
   -- (>>=) = _bind
   -- {-#INLINE (>>=) #-}
   --
   stream >>= f =
     loop stream where
     loop stream0 = case stream0 of
-      Step fstr -> Step (fmap loop fstr)
+      Step fstr k -> Step fstr (loop . k)
       Effect m   -> Effect (liftM loop m)
       Return r  -> f r
   {-# INLINABLE (>>=) #-}       
 
-  fail = lift . fail
+  fail = Effect . fail
```

Comment:
What's this change for? Does it have anything to do with the rest?

Reply:
It may not be necessary, it may have been an attempt to implement `Monad` more directly, rather than relying on `MonadTrans`.